### PR TITLE
Feature: Add Support for Data URI

### PR DESCRIPTION
**Pull request by davleop**
# Feature: add support for data URI

Following after OpenAI's vision docs: https://platform.openai.com/docs/guides/vision

### CODE REVIEW

**Comment by michaelfeil** [comment]
Nice, so you want to implement.

```
def encode_image(image_path):
  with open(image_path, "rb") as image_file:
    return base64.b64encode(image_file.read()).decode('utf-8')
```
And then accept the following?
```
base64_image = encode_image(image_path)
{
"url": f"data:image/jpeg;base64,{base64_image}"
}
```

I like it - please make sure to also add a unit test for this, and add a documentation. If there is an OpenAPI feature that generates `pip install infinity_client`, this would be helpful too.

File: libs/infinity_emb/infinity_emb/transformer/vision/utils.py [hunk 1]
```diff
@@ -49,13 +51,50 @@ async def resolve_from_img_url(
...
+    except Exception as e:
+        raise ImageCorruption(
+            f"error decoding data URI: {e}"
+        )
+
+
+def is_base64_check(s: str):
+    """Regex check to quickly check if string is base64 or not."""
+    pattern = (
+        r"^[A-Za-z0-9+/]{4}([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$"
```
Comment (michaelfeil) [inline libs/infinity_emb/infinity_emb/transformer/vision/utils.py:70]: Is this a performance efficient check for this? Is it needed (doesn't the Mozilla spec say it needs to say base64?

**Comment by michaelfeil** [comment]
@davleop @stikkireddy Added a lot of changes, to make the data uri validation more native to the pydantic validation.

**Comment by stikkireddy** [comment]
@michaelfeil awesome! this is what i was hoping would be changed. you got rid of regex match for base64 format. The mdn guide specifies the format structure and if the base64 key is provided then its assumed that the rest of the string is base64 encoded. If decoding fails we can send that error back to the user.

https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data

**Comment by stikkireddy** [comment]
@michaelfeil 

Last question is there any way to support this from OpenAI embeddings spec? extra_body is typically used to provide additional non spec features to allow the server to handle things. It would be nice to be able to reuse the openai embeddings client for things like:

```python
client = OpenAI(....)
client.embeddings.create(
model="default",
inputs=["https://...", "data:..."],
extra_body={
  "infiniy_image_embeddings": true
}
```

I know its not exactly what openai supports but it would be great to be able to use existing clients to solve this problem if possible. they have built in retries, etc.

File: libs/infinity_emb/infinity_emb/fastapi_schemas/data_uri.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+T = TypeVar("T")
+
+MIMETYPE_REGEX = r"[\w]+\/[\w\-\+\.]+"
+MIMETYPE_REGEX_AUDIO_IMAGE = r"(audio|image)\/[\w\-\+\.]+"
+_MIMETYPE_RE = re.compile("^{}$".format(MIMETYPE_REGEX_AUDIO_IMAGE))
+
+CHARSET_REGEX = r"[\w\-\+\.]+"
+_CHARSET_RE = re.compile("^{}$".format(CHARSET_REGEX))
+
+DATA_URI_REGEX = (
```
Comment (michaelfeil) [inline libs/infinity_emb/infinity_emb/fastapi_schemas/data_uri.py:26]: @stikkireddy Does this regex work?

Comment (stikkireddy) [reply libs/infinity_emb/infinity_emb/fastapi_schemas/data_uri.py:26]: [libs/infinity_emb/infinity_emb/transformer/vision/utils.py](https://github.com/michaelfeil/infinity/pull/385/files/a6175fb056cd39b17402f622f9874214e912fd6e#diff-a47efc277753063b843bf0543182a00f1918f50424d5f9b28877956bef7b4f19)

i assume you will remove that unnecessary code then?

**Comment by davleop** [comment]
> @davleop @stikkireddy Added a lot of changes, to make the data uri validation more native to the pydantic validation.

Thanks for picking up my slack! I haven't had much time to take a look at this again. I kinda just implemented this as a solution I needed at the time and thought it might be good to push it up to main :)

**Pull request merged**